opencog/attention/experimental/HebbianCreationAgent.cc

```diff
@@ -0,0 +1,173 @@
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#include <opencog/util/Config.h>
+#include <opencog/util/algorithm.h>
+
+#include <opencog/atoms/base/Link.h>
+#include <opencog/truthvalue/IndefiniteTruthValue.h>
+#include <opencog/truthvalue/SimpleTruthValue.h>
+#include <opencog/attention/atom_types.h>
+#include <opencog/atomutils/Neighbors.h>
+
+#define DEPRECATED_ATOMSPACE_CALLS
+#include <opencog/atomspace/AtomSpace.h>
+#include <opencog/cogserver/server/CogServer.h>
+
+#include "HebbianCreationAgent.h"
+
+using namespace opencog;
+
+HebbianCreationAgent::HebbianCreationAgent(CogServer& cs) :
+    Agent(cs)
+{
+    // Provide a logger, but disable it initially
+    log = NULL;
+    setLogger(new opencog::Logger("HebbianCreationAgent.log", Logger::FINE, true));
+
+    maxLinkNum = config().get_int("ECAN_MAXLINKS");
+    localToFarLinks = config().get_int("ECAN_LOCAL_FAR_LINK_RATIO");
+
+    as = &_cogserver.getAtomSpace();
+}
+
+HebbianCreationAgent::~HebbianCreationAgent()
+{
+    if (log) delete log;
+}
+
+void HebbianCreationAgent::setLogger(Logger* _log)
+{
+    if (log) delete log;
+    log = _log;
+}
+
+Logger* HebbianCreationAgent::getLogger()
+{
+    return log;
+}
+
+void HebbianCreationAgent::run()
+{
+    Handle source;
+    newAtomsInAV.pop(source);
+
+    //HebbianLinks should not normally enter to the AF boundary since they
+    //should not normally have STI values.The below check will avoid such
+    //Scenarios from happening which could lead to HebbianLink creation
+    //bn atoms containing HebbianLink.
+    if (classserver().isA(source->getType(), HEBBIAN_LINK))
+        return;
+
+    HandleSeq notAttentionalFocus;
+    int afb = as->get_attentional_focus_boundary();
+    as->get_handles_by_AV(back_inserter(notAttentionalFocus),0,afb);
+
+   // Retrieve the atoms in the AttentionalFocus
+    HandleSeq attentionalFocus;
+    as->get_handle_set_in_attentional_focus(back_inserter(attentionalFocus));
+
+    // Exclude the source atom
+    attentionalFocus.erase(std::remove(attentionalFocus.begin(),
+                                       attentionalFocus.end(),
+                                       source), attentionalFocus.end());
+
+    if (source == Handle::UNDEFINED)
+        return;
+
+    // Get the neighboring atoms, where the connecting edge
+    // is an AsymmetricHebbianLink in either direction
+    HandleSeq existingAsSource =
+            get_target_neighbors(source, ASYMMETRIC_HEBBIAN_LINK);
+    HandleSeq existingAsTarget =
+            get_source_neighbors(source, ASYMMETRIC_HEBBIAN_LINK);
+
+    // Get the set differences between the AttentionalFocus
+    // and the sets of existing sources and targets
+    std::sort(existingAsSource.begin(), existingAsSource.end());
+    std::sort(existingAsTarget.begin(), existingAsTarget.end());
+    std::sort(attentionalFocus.begin(), attentionalFocus.end());
+    HandleSeq needToBeSource = set_difference(attentionalFocus,
```

Comment:
I'm not sure, but I think you could save a bunch of cpu time by using std::set instead of std::vector, here.  sets are rb-trees, and are sorted automatically, which allows you to skip the explicit sort step above.